versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/KeyListBuildState.java

```diff
@@ -15,114 +15,155 @@
+          openAddressingHashMap[i] = entry;
+          break;
+        }
+
+        i++;
+        if (i == openAddressingEntries) {
+          i = 0;
+        }
       }
-      addToKeyListEntity(keyListEntry, entrySize);
     }
-  }
 
-  private void finishKeyListEntity() {
-    Hash id = randomHash();
-    newKeyListEntities.add(KeyListEntity.of(id, currentKeyList.build()));
-    newCommitEntry.addKeyListsIds(id);
-  }
+    // Generate all KeyList objects. The first one for the embedded key-list and all other ones
+    // for key-list-entities. Offsets are kept for key-list-entities, because the embedded
+    // key-list's is always the first one.
 
-  private void newKeyListEntity() {
-    currentSize = 0;
-    currentKeyList = ImmutableKeyList.builder();
-  }
+    List<Integer> offsets = new ArrayList<>();
+    List<KeyList> keyLists = new ArrayList<>();
 
-  private void addToKeyListEntity(KeyListEntry keyListEntry, int keyTypeSize) {
-    currentSize += keyTypeSize;
-    currentKeyList.addKeys(keyListEntry);
-  }
+    int bucketSize = 0;
+    int maxBucketSize = maxEmbeddedKeyListSize;
+    ImmutableKeyList.Builder keyListBuilder = ImmutableKeyList.builder();
+    for (int i = 0; i < openAddressingHashMap.length; i++) {
+      KeyListEntry entry = openAddressingHashMap[i];
 
-  private void addToEmbedded(KeyListEntry keyListEntry, int keyTypeSize) {
-    currentSize += keyTypeSize;
-    embeddedBuilder.addKeys(keyListEntry);
-  }
+      if (entry != null) {
+        int entrySize = serializedEntrySize.applyAsInt(entry);
+
+        if (bucketSize + entrySize > maxBucketSize) {
+          maxBucketSize = maxKeyListEntitySize;
+          offsets.add(i);
+          keyLists.add(keyListBuilder.build());
+          keyListBuilder = ImmutableKeyList.builder();
+          bucketSize = 0;
+        }
+
+        bucketSize += entrySize;
+      }
+      keyListBuilder.addKeys(entry);
+    }
+    if (bucketSize > 0) {
+      // Only add the last key-list-entity if it actually contains entries (not all null)
+      keyLists.add(keyListBuilder.build());
+    }
 
-  void finish() {
-    // If there's an "unfinished" KeyListEntity, build it.
-    if (currentKeyList != null) {
-      finishKeyListEntity();
+    // Very rare, but it's possible that there are no keys at all.
+    if (!keyLists.isEmpty()) {
+      newCommitEntry.keyList(keyLists.get(0));
+    } else {
+      newCommitEntry.keyList(KeyList.EMPTY);
     }
-    newCommitEntry.keyList(embeddedBuilder.build());
+    newCommitEntry.keyListLoadFactor(loadFactor);
+
+    List<KeyListEntity> builtEntities =
+        keyLists.stream()
+            .skip(1)
+            .map(keyList -> KeyListEntity.of(randomHash(), keyList))
+            .collect(Collectors.toList());
+
+    if (!builtEntities.isEmpty()) {
+      builtEntities.stream().map(KeyListEntity::getId).forEach(newCommitEntry::addKeyListsIds);
+      newCommitEntry.addAllKeyListEntityOffsets(offsets);
```

Comment:
Another idea, I was considering was to hash over large buckets, but have a sorted list inside each bucket.. I guess that would be a hybrid approach between a hash table of lists and a total open addressing hash table. I suppose the benefit would be that we do not have to keep a separate list on indexes and handling collisions within a bucket is easier (by sorted list). Also buckets can fill up independently without cause a re-hash or extra I/O hops (I suppose it is faster to search within a bucket than load the next hop).

We could even do perfect hashing within each bucket... but a sorted list would be fine, I guess.

So the lookup / insertion would go like this:
* bucket_index = hash % num_buckets
* on add, try fill first bucket up to size limit, rehash if overflow. Only one bucket needs to be re-written unless there's an overflow.
* on read, one bucket load is sufficient (by hash index) if not found, the entry does not exist anywhere.
* on rehash, we undersize buckets to accommodate future inserts.

WDYT?

Reply:
Actually on the fence here between a "pure open-addressing" approach and a "mixed" approach (open-addressing for segmenting + sorted segments).
I've added a new test `AbstractManyKeys.pointContentKeyLookups()` to #4551 - the setting for the test (with ~160 bytes for each key-entry & 16kB entity size) leads to very few "additional fetches". Bumping that 16kB entity size to 64kB yielded no observed. (Hint: can be observed by adding a breakpoint in the line calling `fetchKeyLists()` in `AbstractDatabaseAdapter.fetchValuesHandleOpenAddressingKeyList()` with the condition `round > 0`.)

As I said, not sure yet. Thinking about other related stuff as well:
* Could we return a sorted content-key list to clients?
  Feels like this could be implemented by fetching all key-list-entities and merging the iterators. But that's a minor win.
* How could we eventually implement paging for "get content keys"?
  Feels like that could be easily implemented by with a paging token consisting of the commit-ID + bucket number. But also relatively easy w/ the old and mixed approach: commit-ID + segment number + offset in segment.

One argument against "fixed proportional segments" is the non-deterministic length of keys and potentially bad distribution of hash values, especially w/ non segment counts that are not a power of two. Even if you have e.g. 20000 keys, there can be (equally distributed) segments with no keys, key keys and "too many" keys (exceeding a database limit).